prov/verbs/src/windows/verbs_nd_ov.c

```diff
@@ -313,29 +325,48 @@ void nd_cm_connect_ack(struct nd_event_base *base, DWORD bytes)
 		if (!cm_event)
 			cm_event = nd_allocate_cm_event(
 				id, RDMA_CM_EVENT_ESTABLISHED);
+		break;
 
+	case ND_CM_DISCONNECT:
+	case ND_CM_DISCONNECTED:
+		cm_event = nd_allocate_cm_event(id, RDMA_CM_EVENT_DISCONNECTED);
+		break;
+
+	default:
+		break;
+	}
+
+	VRB_DBG(FI_LOG_EQ, "EQ SET event:%p %p %d\n", cm_event->event.id,
+		cm_event->event.listen_id, cm_event->event.event);
+	dlistfd_insert_tail(&cm_event->entry, &ch_nd->q);
+
+	// The call to NotifyDisconnect needs to happen after we insert the
+	// CM event into the queue in order to avoid a race condition where
+	// a disconnect is virtually immediate and that callback interrupts
+	// this processing leading to an RDMA_CM_EVENT_DISCONNECTED coming
+	// before its associated RDMA_CM_EVENT_ESTABLISHED.
+	switch (event->state) {
```

Comment:
I don't understand what's happening here.  There's a switch statement above this that checks for event->state.  AFAICT, there's no change to event, then we have another switch statement here on event->state.  Why are there 2 switch statements?